### ISSUE TITLE
Batch Processing, Difference in Extracted Features

### ISSUE THREAD

**Issue by enienws** [davidsandberg/facenet#234]
I note something weird when extracting features from 160x160 images.

`# Run forward pass to calculate embeddings`
`feed_dict = { images_placeholder: images, phase_train_placeholder:False }`
`emb = sess.run(embeddings, feed_dict=feed_dict)`

"Code snippet from compare.py at line 57"

In the code snippet above, the extracted features are changing with the number of images fed. For example there are differences between features extracted  by appending one image for SubjectA to "images" list and features extracted by appending two or more images by appending multiple images for SubjectA and/or SubjectB to "images" list. 

I don't understand why such a change occurs when number of requested features are changed.
Sincerely.

**Comment by quanguyhn** [comment]
I also had the same problem and still could not figure out why.

**Comment by lizhiyuanUSTC** [comment]
I guess the model use batch_normalization, but the code didn't save mean and var, you can try a model without batch_normalization.

**Comment by zhuchangjiang** [comment]
I have the same trouble

**Comment by trevorprater** [comment]
Has anyone discovered a fix for this?

**Comment by davidsandberg** [comment]
Yes, I have seen this as well but I haven't had time to look deeper into it. But my feeling is also that it is caused by batch norm.

**Comment by davidsandberg** [comment]
I have been digging a little bit more in this. A script to reproduce the problem can be found in `tmp/debug_batch_size.py`.
`python tmp/debug_batch_size.py /home/david/datasets/lfw/lfw_mtcnnalign_160 /home/david/models/20170512-110547/20170512-110547.pb`
I compared the embeddings between running a forward pass with 1 image and 200 images.
The embeddings differ slightly (difference is in the order of 1e-8). I tried both by importing the meta graph and the frozen graph and the embeddings are identical in the two cases (but still differ when batch size is changed).

**Comment by zo7** [comment]
One kind of sneaky thing that I've noticed this code isn't doing (at least AFAIK) is updating the moving mean and variance for normalization layers. TensorFlow actually now requires that you [call extra update ops when using batch norm layers](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/layers/python/layers/layers.py#L398):
```python
update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
with tf.control_dependencies(update_ops):
  train_op = optimizer.minimize(loss)
```
Perhaps this is one thing that might help the issue?

That doesn't explain why values are changing when the learning phase placeholder is False though. My other thought just briefly sifting through the code is I noticed that many functions default the `is_training` flag in batch norm to a boolean True, it might be possible somewhere that the placeholder isn't being passed through to some layer. (just a thought though, haven't checked deep enough to see if this is true)

**Comment by pavankumarkatakam** [comment]
I have the same problem. I trained my own images and svm to predict images, when im running predictions in video some frames are getting wrong results and when i extract the frames individually and predict them individually the results are correct. So the base problem is different embeddings are getting when in batch or individual. What is the solution for this problem?
